Add Position Specifiers

tcole3 opened the pull request:
Added a new kind of specifier for describing and identifying a position within a text or data stream.
Changes are more extensive, but better isolated from existing Selector and State specifiers
Still need real-world use cases, especially for side-bias.


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
[Preview](https://s3.amazonaws.com/pr-preview/tcole3/publ-loc/master.html) | [Diff](https://s3.amazonaws.com/pr-preview/w3c/publ-loc/9f644a4...tcole3:ad312fb.html)

iherman commented on index.html:
```diff
@@ -243,7 +249,7 @@
             <ol>
                 <li>
-					the <a href="https://s3.amazonaws.com/pr-preview/tcole3/wpub/master.html#dfn-url">URL</a> of the overall resource; we will refer to this as the <a>Source</a>.
+					the <a href="#dfn-url">URL</a> of the overall resource; we will refer to this as the <a>Source</a>.
```
Minor thing: this should actually point to https://w3c.github.io/wpub/index.html#dfn-url, because there is no local definition of a URL. (The original one was indeed a leftover editorial bug.)

iherman commented on index.html:
```diff
@@ -975,32 +981,32 @@
 			<ol>
 				<li>
-					the <a href="https://s3.amazonaws.com/pr-preview/tcole3/wpub/master.html#dfn-url">URL</a> of the overall resource; this is the same <a>Source</a> as used for Selectors (see <a href="#selectors"></a>). 
+					the <a href="#dfn-url">URL</a> of the overall resource; this is the same <a>Source</a> as used for Selectors (see <a href="#selectors"></a>). 
```
Same as above...

iherman commented on index.html:
```diff
@@ -1181,30 +1187,274 @@
+					by the <code>value</code> property (<code>"bias": "before"</code>)
+					or to the character following the position (<code>"bias": "after"</code>). For example, if the text stream 
+					was “abcdefghijklmnopqrstuvwxyz”, the <code>value</code> was 7, and the <code>bias</code> was before,
+					then the position referenced would be the position between "g" and "h" and would be attached to "g".	
```
It should probably use <code>before</code>

iherman commented on index.html:
```diff
@@ -1181,30 +1187,274 @@
 
 			<ul>
-				<li>The fragment uses the <code>selector(…)</code>, respectively the <code>state(…)</code>, functional syntax</li>
+				<li>The fragment uses the <code>selector(…)</code>, <code>position(...)</code>, or <code>state(…)</code>, functional syntax</li>
```
remove `,` before "functional"

iherman commented on index.html:
```diff
@@ -1181,30 +1187,274 @@
+				<li>The fragment uses the <code>selector(…)</code>, <code>position(...)</code>, or <code>state(…)</code>, functional syntax</li>
 				<li>The (comma separated) “parameters” of the functional notation are:
 					<ul>
 						<li>For the keys <code>refinedBy</code>, <code>startSelector</code>, and <code>endSelector</code> the syntax is <code>key=selector(…)</code>, respectively <code>key=state(…)</code> when appropriate, with the value following, recursively, the same syntax as the full fragment;</li>
```
The key=position(...) should be added here as well (which means rewording the sentence a bit...)

iherman commented:
Two additional, editorial notes

* [ ] Right before 1.1 I believe we should emphasize that we introduce a completely different concept, namely Position selector, with two specific Positions. Ie, on the top level we have now three concepts and not two.
* [ ] A slightly more convoluted modification of the example in 5.1 is that George wants to insert some characters in the middle of a word (e.g., for taking care of a spelling mistake) and wants to make sure that the relevant comment or user feedback stays with the part of the word _after_ the position even in case the word is automatically hyphenated. That may be closer to what "bias" means.

iherman commented:
As has been noted many times, _we really need clear use cases for this_. 

My [comment](https://github.com/w3c/publ-loc/pull/15#issuecomment-338150249) on the (otherwise closed) PR #15 remains: implementors have to do something fundamentally different than what they have at their disposal in a browser: while `State` and `Selector` implementations can rely on HTTP facilities and DOM implementations, respectively, implementing positions pushes them in a very different, possibly unchartered territory. _If_ really necessary for strong use cases, we should indeed do this (and this PR is, technically, clear and clean for that!), but I am not convinced of the necessity. 

I am not opposed to merge this PR and discuss afterwards, but a roll-back from it may be complicated, because the changes are spread all over the place. I would really prefer having a decision before a possible merge...

CC @GarthConboy @TzviyaSiegman @llemeurfr @rkwright

iherman commented:
For admin: this PR is relevant for issue #9

tcole3 commented:
I think I've addressed issues raised by Ivan's review of this PR. 

While might still benefit from further wordsmithing, would like to merge as is (especially since there are already subsequent PRs pending). Will do so later today or tomorrow morning, unless there is objection. 

I still have my doubts about side-bias, but I'm convinced there are non-annotation use cases for a clean way to reference position in a text or byte stream (rather than just a selection / fragment of content). Also think Position specifiers will be handy for refinement.

iherman commented:
I will merge it, to make the discussion cleaner for others
